### PR TITLE
Switch to forward-over-reverse mode hvp for autograd backend

### PR DESCRIPTION
**Pull request by nkoep**
This (experimental) PR replaces autograd's hvp implementation with a version based on forward-over-reverse mode autodiff (see https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html?highlight=hvp#hessian-vector-products-using-both-forward-and-reverse-mode). This should generally be more efficient that our handrolled implementation that manually computes the gradient of the directional derivative.

Unfortunately, when development of autograd ceased, there was no JVP support for `numpy.linalg.norm` available, which we use in a few places. Some examples and test cases therefore had to be updated to support this change. I'm currently debating whether or not this limitation is worth it as it likely extends to more primitives than `numpy.linalg.norm`.

### CODE REVIEW

**Comment by nkoep** [comment]
> Hey @nkoep, I've [added](https://github.com/HIPS/autograd/pull/579) the JVP of `autograd.numpy.linalg.norm`. I'm pretty sure all of the differentiable elementwise `ufunc`s in numpy and scipy, and all of the `np.*` namespace has JVPs implemented, but none of `np.linalg.*` does, except for `norm` which I've just implemented.
> 
> Generally speaking it's not difficult to convert a VJP to a JVP, it usually involves replacing broadcasting with summing, and vice versa, and not much else.

Awesome! Thanks a lot for taking the time to implement this, Jamie. I'm still not sure whether moving forward with this change is advisable though since I fear the lack of jvp support for numpy.linalg may be limiting depending on the type of cost function a user is defining. Maybe it's best to hide this behavior behind a flag to the autograd function decorator, so that you would have to opt in explicitly to use reverse-over-forward mode along those lines:
```python
@pymanopt.function.autograd(manifold, reverse_over_forward=True)
def cost(x):
  ...
```
WDYT?

**Comment by nkoep** [comment]
I'm closing this again for now due to the missing jvp support of `autograd.numpy.linalg`.
